File: src/main/java/org/cubewhy/celestial/utils/lunar/LauncherData.java
package org.cubewhy.celestial.utils.lunar;

import com.google.gson.*;
import okhttp3.Response;
import org.cubewhy.celestial.event.impl.CrashReportUploadEvent;
import org.cubewhy.celestial.utils.CrashReportType;
import org.cubewhy.celestial.utils.OSEnum;
import org.cubewhy.celestial.utils.RequestUtils;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URL;
import java.util.*;

public final class LauncherData {
    public final URI api;

    /**
     * Create a LauncherData instance with special API resource
     *
     * @param api Launcher API
     */
    public LauncherData(URI api) {
        super();
        this.api = api;
    }

    /**
     * Create a LauncherData instance with the official Launcher API
     */
    public LauncherData() {
        this(URI.create("https://api.lunarclientprod.com")); // official API
    }

    /**
     * Create a LauncherData instance with special API resource
     *
     * @param api Launcher API
     */
    public LauncherData(String api) {
        this(URI.create(api));
    }

    /**
     * Create a LauncherData instance with special API resource
     *
     * @param api Launcher API
     */
    public LauncherData(@NotNull URL api) throws URISyntaxException {
        this(api.toURI());
    }

    /**
     * Get LunarClient main-class (online)
     *
     * @param json Json of the special LunarClient instance
     * @return main class of the LunarClient instance
     */
    public static String getMainClass(JsonObject json) {
        if (json == null) {
            return "com.moonsworth.lunar.genesis.Genesis";
        }
        return json
                .getAsJsonObject("launchTypeData")
                .get("mainClass").getAsString();
    }

    /**
     * Get ICHOR state
     *
     * @return true (always)
     */
    public static boolean getIchorState(JsonObject json) throws IOException {
        if (json.getAsJsonObject("launchTypeData").has("ichor")) {
            return json
                    .getAsJsonObject("launchTypeData")
                    .get("ichor").getAsBoolean();
        } else {
            return true; // force enable ichor in the latest api version
        }
    }

    /**
     * Get metadata
     *
     * @return Launcher Metadata
     */
    public JsonObject metadata() throws IOException {
        // do request with fake system info
        try (Response response = RequestUtils.get(api + "/launcher/metadata" + "?os=linux" + "&arch=x64" + "&launcher_version=v3.1.3-master").execute()) {
            assert response.code() == 200 : "Code = " + response.code(); // check success
            assert response.body() != null : "ResponseBody was null";
            return JsonParser.parseString(response.body().string()).getAsJsonObject();
        }
    }

    /**
     * Get alert message
     *
     * @param metadata metadata from api
     * @return a map of the alert (title, message)
     * */
    public static Map<String, String> getAlert(JsonObject metadata) {
        if (metadata.has("alert") && !metadata.get("alert").isJsonNull()) {
            JsonObject alert = metadata.getAsJsonObject("alert");
            Map<String, String> map = new HashMap<>();
            map.put("title", alert.get("name").getAsString());
            map.put("message", alert.get("text").getAsString());
            return map;
        } else {
            return null;
        }
    }

    /**
     * Get blog posts
     *
     * @param metadata metadata from api
     * @return a list of blog posts
     */
    public static JsonArray getBlogPosts(JsonObject metadata) {
        return metadata.getAsJsonArray("blogPosts");
    }

    public JsonObject getVersion(String version, String branch, String module) throws IOException {
        JsonObject json = new JsonObject();
        json.addProperty("hwid", "HWID-PUBLIC");
        json.addProperty("installation_id", new UUID(100, 0).toString()); // fake uuid
        json.addProperty("os", Objects.requireNonNull(OSEnum.find(System.getProperty("os.name"))).jsName); // shit js
        json.addProperty("arch", "x64"); // example: x64
        json.addProperty("os_release", "19045.3086");
        json.addProperty("launcher_version", "2.15.1");
        json.addProperty("launch_type", "offline");
        json.addProperty("version", version);
        json.addProperty("branch", branch);
        json.addProperty("module", module);

        try (Response response = RequestUtils.post(api + "/launcher/launch", new Gson().toJson(json)).execute()) {
            assert response.body() != null : "ResponseBody was null";
            return JsonParser.parseString(response.body().string()).getAsJsonObject();
        }
    }

    public static List<String> getDefaultJvmArgs(JsonObject json, File installation) {
        List<String> out = new ArrayList<>();
        for (JsonElement arg : json
                .getAsJsonObject("jre")
                .getAsJsonArray("extraArguments")) {
            if (arg.getAsString().equals("-Djna.boot.library.path=natives")) {
                out.add("-Djna.boot.library.path=\"" + installation + "/" + "natives\"");
                continue;
            }
            out.add(arg.getAsString());
        }
        out.add("-Djava.library.path=\"" + installation + "/" + "natives\"");
        return out;
    }

    /**
     * Get support versions
     *
     * @param metadata LC metadata
     * @return Support versions list
     */
    public static @NotNull Map<String, Object> getSupportVersions(JsonElement metadata) {
        Map<String, Object> map = new HashMap<>();
        List<String> versions = new ArrayList<>();
        JsonArray versionsJson = Objects.requireNonNull(metadata).getAsJsonObject().getAsJsonArray("versions");
        for (JsonElement version : versionsJson) {
            String versionId = version.getAsJsonObject().get("id").getAsString();
            if (version.getAsJsonObject().has("subversions")) {
                for (JsonElement subVersion : version.getAsJsonObject().get("subversions").getAsJsonArray()) {
                    versionId = subVersion.getAsJsonObject().get("id").getAsString();
                }
            }
            if (version.getAsJsonObject().get("default").getAsBoolean()) {
                map.put("default", versionId);
            }
            versions.add(versionId);
        }
        map.put("versions", versions);
        return map;
    }

    /**
     * Get information of a subversion
     *
     * @param version version name
     * @return version json
     */
    public static JsonObject getVersionInMetadata(JsonElement metadata, String version) throws IOException {
        JsonObject metadata1 = metadata.getAsJsonObject();
        for (JsonElement version1 : metadata1.get("versions").getAsJsonArray()) {
            if (version.contains(version1.getAsJsonObject().get("id").getAsString())) {
                for (JsonElement subVersion : version1.getAsJsonObject().get("subversions").getAsJsonArray()) {
                    if (subVersion.getAsJsonObject().get("id").getAsString().equals(version)) {
                        return subVersion.getAsJsonObject();
                    }
                }
            }
        }
        return null;
    }

    /**
     * Get support addons
     *
     * @param metadata LC metadata
     * @param version  Minecraft version
     * @return Module List
     */
    public static @NotNull Map<String, Object> getSupportModules(JsonElement metadata, String version) throws IOException {
        Map<String, Object> map = new HashMap<>();
        List<String> modules = new ArrayList<>();
//        boolean isSubVersion = StringUtils.count(version, '.') >= 2;
        JsonObject version1 = getVersionInMetadata(metadata, version);
        JsonArray modulesJson = Objects.requireNonNull(version1).getAsJsonArray("modules");
        for (JsonElement moduleJson : modulesJson) {
            String id = moduleJson.getAsJsonObject().get("id").getAsString();
            modules.add(id);
            if (moduleJson.getAsJsonObject().get("default").getAsBoolean()) {
                map.put("default", id);
            }
        }
        map.put("modules", modules);
        return map;
    }

    /**
     * Get a list of LunarClient Artifacts
     *
     * @param version artifacts json
     * @return artifact list
     */

    public static Map<String, Map<String, String>> getArtifacts(JsonElement version) throws IOException {
        Map<String, Map<String, String>> out = new HashMap<>();

        JsonObject versionJson = Objects.requireNonNull(version).getAsJsonObject();
        JsonObject launchTypeData = versionJson.getAsJsonObject("launchTypeData");
        JsonArray artifacts = launchTypeData.getAsJsonArray("artifacts");

        for (JsonElement artifact : artifacts) {
            Map<String, String> info = new HashMap<>();
            String key = artifact.getAsJsonObject().get("name").getAsString();
            String url = artifact.getAsJsonObject().get("url").getAsString();
            String sha1 = artifact.getAsJsonObject().get("sha1").getAsString();
            String type = artifact.getAsJsonObject().get("type").getAsString();
            info.put("url", url);
            info.put("sha1", sha1);
            info.put("type", type);
            out.put(key, info);
        }

        return out;
    }

    /**
     * Get the textures' index of LunarClient
     *
     * @param version version info
     * @return textures' index
     */
    @Nullable
    public static Map<String, String> getLunarTexturesIndex(@NotNull JsonElement version) throws IOException {
        JsonObject versionJson = version.getAsJsonObject();
        String indexUrl = versionJson.getAsJsonObject("textures").get("indexUrl").getAsString();
        // get index json
        String baseUrl = getTexturesBaseUrl(version);
        try (Response response = RequestUtils.get(indexUrl).execute()) {
            if (response.body() != null) {
                // parse
                Map<String, String> map = new HashMap<>();
                for (String s : response.body().string().split("\n")) {
                    // filename hashcode
                    map.put(baseUrl + s.split(" ")[0], s.split(" ")[1]);
                }
                return map;
            }
        }
        return null;
    }

    public static String getTexturesBaseUrl(JsonElement version) throws IOException {
        JsonObject versionJson = Objects.requireNonNull(version).getAsJsonObject();
        return versionJson.get("baseUrl").getAsString();
    }


    @NotNull
    public Map<String, String> uploadCrashReport(String trace, @NotNull CrashReportType type, String launchScript) throws IOException {
        Map<String, String> map = new HashMap<>();
        // do request
        JsonObject request = new JsonObject();
        request.addProperty("type", type.jsonName);
        request.addProperty("trace", trace);
        request.addProperty("launchScript", launchScript);
        try (Response response = RequestUtils.post(api + "/launcher/uploadCrashReport", request).execute()) {
            if (response.isSuccessful() && response.body() != null) {
                JsonObject json = JsonParser.parseString(response.body().string()).getAsJsonObject().getAsJsonObject("data");
                String id = json.get("id").getAsString();
                String url = json.get("url").getAsString();
                map.put("id", id);
                map.put("message", json.get("message").getAsString());
                map.put("url", url);
                new CrashReportUploadEvent(id, url).call();
            }
        }
        return map;
    }
}

File: src/main/kotlin/org/cubewhy/celestial/utils/lunar/LauncherData.kt
/*
 * Celestial Launcher <me@lunarclient.top>
 * License under GPLv3
 * Do NOT remove this note if you want to copy this file.
 */

package org.cubewhy.celestial.utils.lunar

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import org.cubewhy.celestial.JSON
import org.cubewhy.celestial.config
import org.cubewhy.celestial.event.impl.CrashReportUploadEvent
import org.cubewhy.celestial.game.AddonMeta
import org.cubewhy.celestial.game.RemoteAddon
import org.cubewhy.celestial.string
import org.cubewhy.celestial.utils.CrashReportType
import org.cubewhy.celestial.utils.OSEnum
import org.cubewhy.celestial.utils.RequestUtils.get
import org.cubewhy.celestial.utils.RequestUtils.post
import org.cubewhy.celestial.utils.arch
import java.net.URI
import java.net.URL
import java.util.*

/**
 * Create a LauncherData instance with the official Launcher API
 */
class LauncherData(val api: URI = URI.create("https://api.lunarclientprod.com")) {
    /**
     * Create a LauncherData instance with special API resource
     *
     * @param api Launcher API
     */
    constructor(api: String) : this(URI.create(api))

    /**
     * Get metadata
     *
     * @return Launcher Metadata
     */

    fun metadata(): LauncherMetadata {
        // do request with fake system info
        // DO NOT REPLACE THE launcher_version FIELD TO config.api.versionSpoof
        get("$api/launcher/metadata?installation_id=${UUID.randomUUID()}&os=${OSEnum.current!!.jsName}&arch=${arch}&launcher_version=114.514.191&branch=master&branch_changed=true&private=true&os_release=114.514").execute()
            .use { response ->
                assert(response.code == 200) {
                    "Code = " + response.code
                }
                assert(response.body != null) { "ResponseBody was null" }
                return JSON.decodeFromString(response.string!!)
            }
    }


    fun getVersion(version: String?, branch: String?, module: String?): GameArtifactInfo {
        val map = mapOf(
            "hwid" to "HWID-PUBLIC",
            "installation_id" to UUID.randomUUID().toString(), // fake uuid
            "os" to OSEnum.current!!.jsName, // shit js
            "arch" to arch, // example: x64
            "os_release" to "19045.3086", // fake os release
            "launcher_version" to config.api.versionSpoof, // fake version
            "launch_type" to "offline",
            "version" to version,
            "branch" to branch,
            "module" to module
        )

        post("$api/launcher/launch", JSON.encodeToString(map)).execute().use { response ->
            assert(response.body != null) { "ResponseBody was null" }
            return JSON.decodeFromString(response.string!!)
        }
    }


    fun uploadCrashReport(trace: String?, type: CrashReportType, launchScript: String?): CrashReportResult? {
        // do request
        val request = mapOf(
            "type" to type.jsonName,
            "trace" to trace,
            "launchScript" to launchScript
        )
        post("$api/launcher/uploadCrashReport", request).execute().use { response ->
            if (response.isSuccessful && response.body != null) {
                val result: CrashReportResult = JSON.decodeFromString(response.string!!)
                CrashReportUploadEvent(result).call()
                return result
            }
        }
        return null
    }

    val plugins: List<RemoteAddon>?
        /**
         * Get the plugin list
         */
        get() {
            val list: MutableList<RemoteAddon> = ArrayList()
            val info = URL("$api/plugins/info")
            var json: Array<PluginInfo>
            try {
                get(info).execute().use { response ->
                    assert(response.body != null)
                    json = JSON.decodeFromString(response.string!!)
                }
            } catch (e: Exception) {
                return null // official api
            }
            for (plugin in json) {
                list.add(
                    RemoteAddon(
                        plugin.name,
                        URL(api.toString() + plugin.downloadLink),
                        plugin.sha1,
                        plugin.category,
                        plugin.meta
                    )
                )
            }
            return list
        }

    companion object {
        /**
         * Get LunarClient main-class (online)
         *
         * @param json Json of the special LunarClient instance
         * @return main class of the LunarClient instance
         */
        fun getMainClass(json: GameArtifactInfo? = null): String =
            json?.launchTypeData?.mainClass ?: "com.moonsworth.lunar.genesis.Genesis"

        fun check(api: String): Boolean {
            try {
                get(api).execute()
                return true
            } catch (e: Exception) {
                return false
            }
        }

        fun getDefaultJvmArgs(json: GameArtifactInfo): List<String> {
            val out: MutableList<String> = ArrayList()
            for (arg in json.jre.extraArguments) {
                // block sentry
                if (arg.startsWith("-Dichor.filteredGenesisSentries")) {
                    out.add("-Dichor.filteredGenesisSentries=.*") // block sentry
                    continue
                }
                out.add(arg)
            }
            out.add("-Djava.library.path=natives")
            return out
        }

        /**
         * Get support versions
         *
         * @param metadata LC metadata
         * @return Support versions list
         */
        fun getSupportVersions(metadata: LauncherMetadata): Map<String, Any> {
            val map: MutableMap<String, Any> = HashMap()
            val versions: MutableList<String> = ArrayList()
            for (version in metadata.versions) {
                var versionId: String
                for (subVersion in version.subversions) {
                    versionId = subVersion.id
                    if (version.default) {
                        map["default"] = versionId
                    }
                    versions.add(versionId)
                }
            }
            map["versions"] = versions
            return map
        }

        /**
         * Get information of a subversion
         *
         * @param version version name
         * @return version json
         */
        fun getVersionInMetadata(metadata: LauncherMetadata, version: String): LunarSubVersion? {
            for (version1 in metadata.versions) {
                if (version.contains(version1.id))
                    for (subVersion in version1.subversions)
                        if (subVersion.id == version) return subVersion
            }
            return null
        }

        /**
         * Get support addons
         *
         * @param metadata LC metadata
         * @param version  Minecraft version
         * @return Module List
         */
        fun getSupportModules(metadata: LauncherMetadata, version: String): Map<String, Any> {
            val map: MutableMap<String, Any> = HashMap()
            val modules: MutableList<String> = ArrayList()
            val version1 = getVersionInMetadata(metadata, version)
            val modulesJson = version1!!.modules
            for (moduleJson in modulesJson) {
                val id = moduleJson.id
                modules.add(id)
                if (moduleJson.default) {
                    map["default"] = id
                }
            }
            map["modules"] = modules
            return map
        }

        /**
         * Get the textures' index of LunarClient
         *
         * @param version version info
         * @return textures' index
         */

        fun getLunarTexturesIndex(version: GameArtifactInfo): Map<String, String>? {
            val indexUrl = version.textures.indexUrl
            // get index json
            val baseUrl = version.textures.baseUrl
            return getIndex(baseUrl, indexUrl)
        }

        fun getLunarUiAssetsIndex(version: GameArtifactInfo): Map<String, String> {
            if (version.ui == null) return emptyMap() // there's no ui files for old LC
            return getIndex(version.ui.assets.baseUrl, version.ui.assets.indexUrl)
        }

        private fun getIndex(baseUrl: String, indexUrl: String): Map<String, String> {
            get(indexUrl).execute().use { response ->
                if (response.body != null) {
                    // parse
                    val map: MutableMap<String, String> = HashMap()
                    for (s in response.body!!.string().split("\n".toRegex()).dropLastWhile { it.isEmpty() }
                        .toTypedArray()) {
                        // filename hashcode
                        map[s.split(" ".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]] =
                            baseUrl + s.split(" ".toRegex()).dropLastWhile { it.isEmpty() }
                                .toTypedArray()[1]
                    }
                    return map
                }
            }
            return emptyMap()
        }
    }
}

// Entities

@Serializable
data class LunarVersion(
    val id: String,
    val default: Boolean,
    val subversions: List<LunarSubVersion>
)

@Serializable
data class LunarSubVersion(
    val id: String,
    val default: Boolean,
    val modules: List<LunarModule>
)

@Serializable
data class GameArtifactInfo(
    val launchTypeData: LaunchTypeData,
    val textures: Textures,
    val jre: RuntimeInfo,
    val ui: UiInfo? = null,
    val canaryToken: String? = ""
) {
    @Serializable
    data class RuntimeInfo(
        val extraArguments: List<String>
    )

    @Serializable
    data class Artifact(
        val name: String,
        val sha1: String,
        val url: String,
        val type: ArtifactType,
    ) {
        enum class ArtifactType {
            CLASS_PATH,
            EXTERNAL_FILE,
            NATIVES,
            JAVAAGENT // LCCN API, not support yet
        }
    }

    @Serializable
    data class LaunchTypeData(
        val artifacts: List<Artifact>,
        val mainClass: String,
        val ichor: Boolean = true
    )

    @Serializable
    data class Textures(
        val indexUrl: String,
        val indexSha1: String,
        val baseUrl: String
    )

    @Serializable
    data class UiInfo(
        val sourceUrl: String,
        val sourceSha1: String,
        val assets: UiAssets
    )

    @Serializable
    data class UiAssets(
        val baseUrl: String,
        val indexUrl: String,
        val indexSha1: String
    )
}

@Serializable
data class LunarModule(
    val id: String,
    val default: Boolean,
)

@Serializable
data class Blogpost(
    val title: String,
    val excerpt: String? = null, // only available on LCCN API
    val image: String,
    val link: String,
    val author: String? = null, // moonsworth removed this in v3.0.0
    @SerialName("button_text")
    val buttonText: String? = null,
    val type: ButtonType? = ButtonType.OPEN_LINK
) {
    enum class ButtonType {
        OPEN_LINK,
        CHANGE_API
    }
}

@Serializable
data class Alert(
    val name: String? = "Alert",
    val text: String? = null,
    val link: String? = null
)

@Serializable
data class LauncherMetadata(
    val versions: List<LunarVersion>,
    @SerialName("blogPosts")
    val blogposts: List<Blogpost> = emptyList(),
    val alert: Alert? = null
)

@Serializable
data class PluginInfo(
    val name: String,
    val sha1: String,
    val downloadLink: String,
    val category: RemoteAddon.Category,
    val meta: AddonMeta? = null
)

@Serializable
data class CrashReportResult(
    val id: String,
    val url: String,
    val message: String
)
